sdk/core/Azure.Core/src/Operation.cs

```diff
@@ -25,8 +25,33 @@ public abstract class Operation
         /// <param name="rehydrationToken">The rehydration token.</param>
         /// <param name="options">The client options.</param>
         /// <returns>The long-running operation.</returns>
-        public static Operation<T> Rehydrate<T>(HttpPipeline pipeline, RehydrationToken? rehydrationToken, ClientOptions? options = null) where T : IPersistableModel<T>
-            => new RehydrationOperation<T>(pipeline, rehydrationToken, options);
+        public static Operation<T> Rehydrate<T>(HttpPipeline pipeline, RehydrationToken rehydrationToken, ClientOptions? options = null) where T : IPersistableModel<T>
+        {
+            Argument.AssertNotNull(pipeline, nameof(pipeline));
+            Argument.AssertNotNull(rehydrationToken, nameof(rehydrationToken));
+
+            IOperationSource<T> source = new GenericOperationSource<T>();
+            var nextLinkOperation = (NextLinkOperationImplementation)NextLinkOperationImplementation.Create(pipeline, rehydrationToken);
+            var operation = NextLinkOperationImplementation.Create(source, nextLinkOperation);
+            var operationState = operation.UpdateStateAsync(false, default).EnsureCompleted();
```

Comment:
Should these call the sync `UpdateState` instead of doing sync-over-async?

Reply:
Ah, I see, NextLinkOperationImplementation does not have a sync `UpdateState`.

Could you please add a sync `UpdateState` method to NextLinkOperationImplemenation?  You can follow the example of how to switch on the `async` bool value to write the code once that is implemented in ClientRetryPolicy [here](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs#L42-L49).

My specific concern is that any time you call `GetAwaiter().GetResult()` (which happens in the call to EnsureCompleted) on a Task or ValueTask, if the task has not completed asynchronously, it could cause deadlocks if there aren't a lot of threads in the threadpool.

Following the pattern in ClientRetryPolicy should let you achieve this without risking deadlocks.

Reply:
sounds good, will add it in a separate PR for NextLinkOperationImplementation.

Reply:
sounds good, will add it in a separate PR for `NextLinkOperationImplementation`.

Reply:
I had another look at this, this is way beyond the scope of the current PR.
Everything works with the current async operation pattern,  if we want to add a sync operation.
We have the internal [IOperation.UpdateStateAsync](https://github.com/Azure/azure-sdk-for-net/blob/c3105fa3c4fa4071918def584f010fc1cd37e7bc/sdk/core/Azure.Core/src/Shared/OperationInternal.cs#L174) in place, we need to split this into async and sync methods.
There are two concerns doing this:
- We want to eventually remove this internal interface, updating it is another technical debt
- All operation implementations are based on this, we need to update them all

Overall, this PR doesn't make this worse, just follow the existing pattern we have. Can we do it in a later PR to fix this all? Preferably, we need to come up with a better design to remove the internal interface `IOperation` to fix this properly.

cc @m-nash @KrzysztofCwalina @tg-msft

Reply:
@annelo-msft I had a look at the implementation in `ClientRetryPolicy` [here](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs#L42-L49), the async methods calls `ProcessSyncOrAsync(message, pipeline, currentIndex, async: false).EnsureCompleted()`, which is same as `operation.UpdateStateAsync(false, default).EnsureCompleted()`.

If we want to properly split async and sync methods, I think we need to split [NextLinkOperationImplementation.GetOperationAsync](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/src/Shared/NextLinkOperationImplementation.cs#L266) into async and sync methods. And then, provide sync and async methods all the ways to the user, as described in above comment.

Reply:
You're right that the sync method calls that.  The difference is that in ClientRetryPolicy, the sync method never calls `await` and blocks on getting the result back.  That is what we need to do here to avoid sync-over-async.  Let me know if you need assistance implementing this -- I can provide a sample or make a PR showing you how it can be done in not too many lines of code.

Reply:
> sounds good, will add it in a separate PR for NextLinkOperationImplementation.

I would strongly prefer that we not merge a PR that doesn't follow our patterns and opens the door to performance issues and/or deadlocks when using the API.   Please include the requested change here, or help me understand why you feel that it cannot be.    

To be clear, this is a ship stopper under normal circumstances.  If there's an urgent business need, please let me know and we can discuss a short-term exemption with the architects.

Reply:
> > sounds good, will add it in a separate PR for NextLinkOperationImplementation.
> 
> I would strongly prefer that we not merge a PR that doesn't follow our patterns and opens the door to performance issues and/or deadlocks when using the API. Please include the requested change here, or help me understand why you feel that it cannot be.
> 
> To be clear, this is a ship stopper under normal circumstances. If there's an urgent business need, please let me know and we can discuss a short-term exemption with the architects.

@jsquire I was trying to raise a separate PR to mitigate the issue for `NextLinkOperationImpelementation`, I didn't mean to merge this PR without resolving ths issue.
But unfortunately that can't be done easily.

Reply:
@annelo-msft This is resolved in https://github.com/Azure/azure-sdk-for-net/pull/43060, please help review it.

Reply:
Looks good - I approved https://github.com/Azure/azure-sdk-for-net/pull/43060, and thanks for doing that!

One last tiny (non-blocking) nit here -- would you mind using the parameter name in the method calls to UpdateStateAsync, since the method name says it's an async method?  e.g. `UpdateStateAsync(async: false, default).EnsureCompleted()` and `UpdateStateAsync(async: true, default).ConfigureAwait(false)`, just so it's clear what's happening from the call-site?  Many thanks!

Reply:
updated